File: bin/spice/cmd/eval.go
package cmd

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"os"

	"github.com/spf13/cobra"
	"github.com/spiceai/spiceai/bin/spice/pkg/api"
	"github.com/spiceai/spiceai/bin/spice/pkg/context"
	"github.com/spiceai/spiceai/bin/spice/pkg/util"
)

type EvalRequest struct {
	Model string `json:"model"`
}

type EvalResponse struct {
	ID        string             `json:"id"`
	CreatedAt string             `json:"created_at"`
	Dataset   string             `json:"dataset"`
	Model     string             `json:"model"`
	Status    string             `json:"status"`
	Scorers   []string           `json:"scorers"`
	Metrics   map[string]float64 `json:"metrics"`
}

type EvalResult struct {
	Input  string `json:"input"`
	Output string `json:"output"`
	Actual string `json:"actual"`
}

var evalCmd = &cobra.Command{
	Use:   "eval [eval-name]",
	Short: "Run model evaluation",
	Example: `
spice eval tetris --model "my_model"`,
	Run: func(cmd *cobra.Command, args []string) {
		if len(args) < 1 {
			slog.Error("eval name is required")
			return
		}
		evalName := args[0]

		model, err := cmd.Flags().GetString("model")
		if err != nil || model == "" {
			slog.Error("model is required")
			return
		}

		request := EvalRequest{Model: model}
		body, err := json.Marshal(request)
		if err != nil {
			slog.Error("marshaling request", "error", err)
			return
		}

		postBody := string(body)

		rtcontext, err := context.FromFlags(cmd.Flags())
		if err != nil {
			slog.Error("failed to initialize runtime context", "error", err)
			os.Exit(1)
		}
		response, err := api.PostRuntime[[]EvalResponse](rtcontext, fmt.Sprintf("/v1/evals/%s", evalName), &postBody)
		if err != nil {
			slog.Error("running evaluation", "error", err)
			return
		}

		table := make([]interface{}, len(response))
		for i, r := range response {
			table[i] = r
		}
		util.WriteTable(table)
	},
}

func init() {
	evalCmd.Flags().String("model", "", "Model to evaluate")
	RootCmd.AddCommand(evalCmd)
}

File: bin/spice-cli/src/commands/eval.rs
/*
Copyright 2024-2026 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

//! `spice eval` command - Run model evaluation.

use crate::RuntimeContext;
use crate::error::{InvalidArgumentSnafu, InvalidResponseSnafu, Result};
use crate::output::TableOutput;
use clap::Args;
use serde::{Deserialize, Serialize};
use snafu::ensure;
use std::collections::HashMap;

/// Arguments for the `eval` command.
#[derive(Args, Debug)]
pub struct EvalArgs {
    /// Name of the eval to run
    pub eval_name: String,

    /// Model to evaluate
    #[arg(long, required = true)]
    pub model: String,
}

/// Request body for eval API.
#[derive(Debug, Serialize)]
struct EvalRequest {
    model: String,
}

/// Response from eval API.
#[derive(Debug, Deserialize)]
struct EvalResponse {
    id: String,
    created_at: String,
    dataset: String,
    model: String,
    status: String,
    scorers: Vec<String>,
    metrics: HashMap<String, f64>,
}

/// Execute the `eval` command.
///
/// # Errors
///
/// Returns an error if the eval name is empty, model is empty, or the API request fails.
pub async fn execute(ctx: &RuntimeContext, args: &EvalArgs) -> Result<()> {
    ensure!(
        !args.eval_name.is_empty(),
        InvalidArgumentSnafu {
            message: "eval name is required"
        }
    );

    ensure!(
        !args.model.is_empty(),
        InvalidArgumentSnafu {
            message: "model is required"
        }
    );

    let request = EvalRequest {
        model: args.model.clone(),
    };

    let response = ctx
        .post_json(&format!("/v1/evals/{}", args.eval_name), &request)
        .await?;

    if !response.status().is_success() {
        let status = response.status();
        let text = response.text().await.unwrap_or_default();
        return Err(InvalidResponseSnafu {
            message: format!("eval request failed with status {status}: {text}"),
        }
        .build());
    }

    let results: Vec<EvalResponse> = response.json().await.map_err(|e| {
        InvalidResponseSnafu {
            message: format!("Failed to parse eval response: {e}"),
        }
        .build()
    })?;

    if results.is_empty() {
        println!("No evaluation results.");
        return Ok(());
    }

    // Display results in a table
    let mut table = TableOutput::new(vec![
        "ID",
        "Created At",
        "Dataset",
        "Model",
        "Status",
        "Scorers",
        "Metrics",
    ]);

    for result in results {
        let scorers = result.scorers.join(", ");
        let metrics = result
            .metrics
            .iter()
            .map(|(k, v)| format!("{k}: {v:.4}"))
            .collect::<Vec<_>>()
            .join(", ");

        table.add_row(vec![
            result.id,
            result.created_at,
            result.dataset,
            result.model,
            result.status,
            scorers,
            metrics,
        ]);
    }

    println!("{table}");

    Ok(())
}
